htlcswitch/link_test.go

```diff
@@ -6894,3 +6895,216 @@ func TestChannelLinkShortFailureRelay(t *testing.T) {
 	default:
 	}
 }
+
+func TestLinkDrainApiDirectionIsolation(t *testing.T) {
+	aliceLink, _, _, _, _, _ :=
+		newSingleLinkTestHarness(
+			t, 5*btcutil.SatoshiPerBitcoin,
+			1*btcutil.SatoshiPerBitcoin,
+		)
+
+	for i := 0; i < 10; i++ {
+		if prand.Uint64()%2 == 0 {
+			//nolint:errcheck
+			aliceLink.EnableAdds(Outgoing)
+		} else {
+			//nolint:errcheck
+			aliceLink.DisableAdds(Outgoing)
+		}
+		require.False(t, aliceLink.IsDraining(Incoming))
+	}
+
+	//nolint:errcheck
+	aliceLink.EnableAdds(Outgoing)
+
+	for i := 0; i < 10; i++ {
+		if prand.Uint64()%2 == 0 {
+			//nolint:errcheck
+			aliceLink.EnableAdds(Incoming)
+		} else {
+			//nolint:errcheck
+			aliceLink.DisableAdds(Incoming)
+		}
+		require.False(t, aliceLink.IsDraining(Outgoing))
+	}
+}
+
+func TestLinkDrainApiGateStateIdempotence(t *testing.T) {
+	aliceLink, _, _, _, _, _ :=
+		newSingleLinkTestHarness(
+			t, 5*btcutil.SatoshiPerBitcoin,
+			1*btcutil.SatoshiPerBitcoin,
+		)
+
+	for _, dir := range []LinkDirection{Incoming, Outgoing} {
+		require.Nil(t, aliceLink.DisableAdds(dir))
+		require.True(t, aliceLink.IsDraining(dir))
+
+		require.NotNil(t, aliceLink.DisableAdds(dir))
+		require.True(t, aliceLink.IsDraining(dir))
+
+		require.Nil(t, aliceLink.EnableAdds(dir))
+		require.False(t, aliceLink.IsDraining(dir))
+
+		require.NotNil(t, aliceLink.EnableAdds(dir))
+		require.False(t, aliceLink.IsDraining(dir))
+	}
+}
+
+func TestLinkOutgoingCommitHooksCalled(t *testing.T) {
+	aliceLink, _, batchTicker, start, _, err :=
+		newSingleLinkTestHarness(
+			t, 5*btcutil.SatoshiPerBitcoin,
+			btcutil.SatoshiPerBitcoin,
+		)
+	if err != nil {
+		t.Fatal()
+	}
+
+	if err = start(); err != nil {
+		t.Fatal("could not start link")
+	}
+
+	hookCalled := make(chan struct{})
+	aliceLink.OnCommitOnce(Outgoing, func() {
+		close(hookCalled)
+	})
+
+	select {
+	case <-hookCalled:
+		t.Fatal("hook called prematurely")
+	case <-time.NewTimer(time.Second).C:
+	}
+
+	batchTicker <- time.Now()
+
+	select {
+	case <-hookCalled:
+	case <-time.NewTimer(time.Second).C:
+		t.Fatal("hook not called")
+	}
+}
+
+func TestLinkCommitHooksRemovable(t *testing.T) {
+	aliceLink, _, batchTicker, start, _, err :=
+		newSingleLinkTestHarness(
+			t, 5*btcutil.SatoshiPerBitcoin,
+			btcutil.SatoshiPerBitcoin,
+		)
+	if err != nil {
+		t.Fatal()
+	}
+
+	if err = start(); err != nil {
+		t.Fatal("could not start link")
+	}
+
+	hookCalled := make(chan struct{})
+	hid := aliceLink.OnCommitOnce(Outgoing, func() {
+		close(hookCalled)
+	}).UnsafeFromSome()
```

Comment:
Ah ok I see the usage here now, so far relegated to just test code.

Reply:
Or it could fail the test early if `None` at that point?

Reply:
The ergonomics were worse that way.

Reply:
It's not just about the assertion. It's about getting our hands on the value inside Some as well if it's not None so we can continue with the rest of the test.